service/controller/v26/resource/endpoint/current.go

```diff
@@ -113,6 +113,20 @@ func (r *Resource) GetCurrentState(ctx context.Context, obj interface{}) (interf
 				return nil, nil
 			}
 		}
+	} else if key.IsPodDeleted(pod) {
+		_, err := r.k8sClient.CoreV1().Pods(pod.GetNamespace()).Get(pod.GetName(), metav1.GetOptions{})
+		if errors.IsNotFound(err) {
+			// In case we reconcile a pod we cannot find anymore this means the
+			// informer's watch event is outdated and the pod got already deleted in
+			// the Kubernetes API. We want to cancel reconciliation to prevent anymore
+			// endpoint IPs of deleted pods being added to services.
+			r.logger.LogCtx(ctx, "level", "debug", "message", "did not find current version of the reconciled pod in the Kubernetes API")
+			r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource")
+			resourcecanceledcontext.SetCanceled(ctx)
+			return nil, nil
+		} else if err != nil {
+			return nil, microerror.Mask(err)
+		}
```

Comment:
What's the general problem with adding this IP? Can it be that update even is processed after delete event? I think they should come in order.

Reply:
This is the point. According one Marcel's argument event might come not in order

Reply:
We observed the problem during KVM updates with many nodes now.
For some reason an endpoint IP is still there even though the pod has been deleted. My idea here was to ensure we don't process an update event after the pod is already gone.

I can't verify that this is happening, but it might be an explanation.

Reply:
> The reasoning behind this check is that even though an object might already be deleted, we could receive an update event for it.

Yes that can happen. But we should receive delete event afterwards and cleanup the mess.

Reply:
@kopiczko If that were the case, then the behaviour we are seeing should be impossible IMO.

Or I simply do not understand how it could happen from the code. Therefor my assumption was that `delete event` -> `update event` -> `actual deletion on k8s` is a possibility. 
That would leave an `update event` as the last event we receive.

Reply:
You cannot get an update event for a deleted object.

Reply:
Why not? The only uncertainty to me is if it can be the last event you get.
Example:
1. Pod has a deletion timestamp and finalizer set. `deletion event` was sent.
2. Number of finalizers is changed -> last one removed
3. Change of finalizers causes `update event`
4. `delete event` should be emitted - this is the point of contention if it will always be last and always be emitted.
5. Resource is fully deleted on k8s side.

The event queue for the operator might process 3. but 5. might have already happened on k8s side as the queue is backed up. Therefore we process an update event for a deleted object.

Reply:
The change of a finalizer does not cause an update event on a deleted object. It is then treated as delete event. The code doing this is in https://github.com/giantswarm/operatorkit/blob/e8903a09b84da1b764da51ed34659e490b66a385/informer/informer.go#L397-L412. 

When you have an idea here, try to write a test for it. Implementing something without verifying the implementation in a reliable way is the worst thing we can do.

Reply:
We also have an integration test for the finalizer management. See https://github.com/giantswarm/operatorkit/blob/master/controller/integration/test/reconciliation. When this is missing something, let's improve this as well.

Reply:
I'm pretty certain removing finalizer creates another delete event. But also operatorkit drops that extra event.